### PR TITLE
Add autoconfig for HOLTEK's Twin Shock

### PR DESCRIPTION
**Pull request by requeijaum**
image

Another DualShock rip-off for PC.
Originally in Brazilian market since 2004 under "Clone" OEM branding.
I need to test the analog mode, though.

On my controller: the analog mode PCB contact is broken. I might solder a push-button later.
The analog axis from left and right thumb are simple guesses from my usage with a Windows driver... long time ago.

-----

Edit: the D-pad usually broke a lot on these controllers. There was a recall but that didn't resolve anything. I still have mine and it's plastic is very strong.

Also: mind that the buttons are numbered starting from 0 and not 1 - as printed WRONGLY on the controller.

### CODE REVIEW

**Comment by hizzlekizzle** [comment]
hey, thanks for this. is there a reason the button labels are all commented out?

**Comment by requeijaum** [comment]
I preferred not to mess with the buttons labels since users can make a mess of it.
That's all because of the buttons markings wearing out and there is an offset of "1" between a marking and the real button (eg. button 0 is marked as 1).

**Comment by offalynne** [comment]
the 0/1 difference isn’t a mistake; windows enumerates buttons in a “human readable” way in user-facing interfaces, ie 1-n instead of 0-n.

File: HOLTEK_Twin-Shock.cfg [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+input_l_y_plus_axis = "+1"
+input_l_y_minus_axis = "-1"
+input_r_x_plus_axis = "+3"
+input_r_x_minus_axis = "-3"
+input_r_y_plus_axis = "+4"
+input_r_y_minus_axis = "-4"
+input_menu_toggle_btn = "8"
+input_b_btn_label = "3"
+input_y_btn_label = "4"
+#input_select_btn_label = "Select"
```
Comment (RobLoach) [inline HOLTEK_Twin-Shock.cfg:32]: No need to comment these out, as mentioned above :wink:

**Review by RobLoach — CHANGES_REQUESTED** [review]

File: HOLTEK_Twin-Shock.cfg [hunk 2]
```diff
@@ -0,0 +1,54 @@
+input_device = "HOLTEK Twin Shock"
```
Comment (RobLoach) [inline HOLTEK_Twin-Shock.cfg:1]: Would be good to have a `input_device_display_name` too.

**Comment by offalynne** [comment]
@requeijaum how did you get the vendor and product fields here? did you author the cfg manually ? the autoconfig format uses decimal and these look to be hex values

**Comment by requeijaum** [comment]
I used macOS System Profiler, Linux's lsusb and Windows 10's Device Manager.
VID and PID are hex values. I didn't know about autoconfig's decimal format.


Em sáb., 13 de mar. de 2021 às 19:05, alynne o. ***@***.***>
escreveu:

**Comment by requeijaum** [comment]
Hey, @RobLoach 
What do you think? Are the USB VID and PID values OK ?
I will also approve your requested changes tomorrow.

**Comment by offalynne** [comment]
assuming rob doesn’t have the hardware on hand to check. they are indeed supposed to be decimal, check other configs to compare

**Comment by RobLoach** [comment]
Merging this, going to fix it up after the merge.

**Pull request merged**